File: java/Binary_Search/Day16/SearchMatrix1.java
import java.util.*;

public class SearchMatrix1 {
    public boolean searchMatrix(int[][] arr, int n, int m, int target) {
        for (int i = 0; i < n; i++) {
            // for (int j = 0; j < m; j++) {
            //     if (arr[i][j] == target) return true;
            // }
            
            if (arr[i][0] <= target && target <= arr[i][m - 1]) {
                int low = 0;
                int high = m - 1;

                // Binary search on the row
                while (low <= high) {
                    int mid = low + (high - low) / 2;

                    if (arr[i][mid] == target) return true;
                    else if (arr[i][mid] < target) low = mid + 1;
                    else high = mid - 1;
                }
            }
        }
        return false;
    }





    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of rows: ");
        int n = sc.nextInt();

        System.out.print("Enter number of columns: ");
        int m = sc.nextInt();

        int[][] arr = new int[n][m];

        System.out.println("Enter matrix elements:");
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                arr[i][j] = sc.nextInt();
            }
        }

        System.out.print("Enter target: ");
        int target = sc.nextInt();

        SearchMatrix1 sm = new SearchMatrix1();

        boolean found = sm.searchMatrix(arr, n, m, target);
        System.out.println("Target found: " + found);


        sc.close();
    }
}

File: python/Binary_Search/day16/searchMatrix1.py
# def search_matrix(arr, target):
#     for row in arr:
#         for val in row:
#             if val == target:
#                 return True
#     return False


def search_matrix(matrix, target):
    n = len(matrix)
    m = len(matrix[0])

    low, high = 0, n * m - 1

    while low <= high:
        mid = (low + high) // 2
        row = mid // m
        col = mid % m

        if matrix[row][col] == target:
            return True
        elif matrix[row][col] < target:
            low = mid + 1
        else:
            high = mid - 1

    return False







if __name__ == "__main__":
    n = int(input("Enter number of rows: "))
    m = int(input("Enter number of columns: "))

    arr = []
    print("Enter matrix elements row by row:")
    for _ in range(n):
        row = list(map(int, input().split()))
        arr.append(row)

    target = int(input("Enter target: "))

    ans = search_matrix(arr, target)
    print("Target found:", ans)